File: Data Structures & Algorithms/max-water-container/submission-1.js
class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */
    maxArea(heights) {
        let left=0;
        let right=heights.length-1;
        let maxArea=0;
        while(left<right){
            let area=Math.min(heights[left],heights[right])*(right-left);
            maxArea=Math.max(maxArea,area);
            if(heights[right]< heights[left]){
                right--;
            }else{
                left++;
            }
        }
        return maxArea;
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isBalanced(TreeNode root) {
        Stack<TreeNode> stack=new Stack<>();
        TreeNode node = root, last=null;
        Map<TreeNode,Integer> depths = new HashMap<>();

        while(!stack.isEmpty() || node!=null){
            if(node!=null){
                stack.push(node);
                node=node.left;
            }else{
                node=stack.peek();
                if(node.right==null || last==node.right){
                    stack.pop();
                    int left=depths.getOrDefault(node.left,0);
                    int right=depths.getOrDefault(node.right,0);
                    if(Math.abs(left-right)>1) return false;
                    depths.put(node,1+Math.max(left,right));
                    last=node;
                    node=null;
                }else{
                    node=node.right;
                }
            }
        }
        return true;
    }
}
